src/extension/src/ActionHandler.py

```diff
@@ -58,27 +58,40 @@ def determine_operation(self, command):
         except KeyError as e:
             raise e
 
+    def setup(self, action, log_message):
+        self.setup_file_logger(action)
+        self.setup_telemetry()
+        self.logger.log(log_message)
+
     def setup_file_logger(self, action):
         if self.file_logger is not None or self.stdout_file_mirror is not None:
             self.logger.log_error("Log file handles from the previous operation were not closed correctly. Closing them and initializing new ones for this operation.")
-            self.close_file_logger()
+            self.tear_down()
 
         log_file_name = datetime.datetime.utcnow().strftime("%Y-%m-%d_%H-%M-%S") + "_" + str(action)
         self.file_logger = self.utility.create_log_file(self.ext_env_handler.log_folder, log_file_name)
         if self.file_logger is not None:
             self.logger.file_logger = self.file_logger
             self.stdout_file_mirror = StdOutFileMirror(self.file_logger)
 
-    def close_file_logger(self):
+    def tear_down(self):
         if self.stdout_file_mirror is not None:
             self.stdout_file_mirror.stop()
         if self.file_logger is not None:
             self.file_logger.close()
 
+    def setup_telemetry(self):
+        # check if events folder exists, if it does init telemetry, if events folder does not exist, log that telemetry is not supported by agent since events folder does not exist
+        events_folder = self.ext_env_handler.events_folder
+        if events_folder is None or not os.path.exists(events_folder):
+            self.logger.log_warning("Telemetry is not supported by Linux Guest Agent as no events folder location specified in Handler Environment")
+        else:
+            self.logger.log("Telemetry is supported by Linux Guest Agent. Extension will send messages to telemetry")
+            self.logger.telemetry_writer.events_folder_path = events_folder
+
```

Comment:
Something important from the full review:
Telemetry is not best effort.. The design tenet here is:

- Does the agent support telemetry? No = don't allow execution to proceed. Write the error to CRP through the status file asking the customer to update their agent.
- If the agent supports telemetry, but telemetry is somehow not working (in a way we can detect), we need to stop execution again with the same process.

Important: tests on the final code too must happen against an older agent version for graceful execution stop with clear message going up to the API through status. Please plan for this right now as new VMs created are expected to all get the new agent version later.

Reply:
Important for public preview:
events_folder = self.ext_env_handler.events_folder is not sufficient..

It must check events_folder, then events_folder_preview and then finally fail.

Reply:
Few concerns here:

- Does the agent support telemetry? No = don't allow execution to proceed. Write the error to CRP through the status file asking the customer to update their agent
1. setup_telemetry is defined within ActionHandler, and used within setup for each handler action (i.e. Install, Enable, Update, etc). 
2. status file is based on seq no. We read/fetch/get seq no only within Enable command. 
3. So at this point in code, handler does not have the seq no and thereby no <seq no>.status file.
4. So the question for us now is, do we check for telemetry on all handler commands or only for enable? Because enable is the only action which writes to status file right now. 
5. If we want to write to status file then telemetry support needs to be checked from within enable handler

- If the agent supports telemetry, but telemetry is somehow not working (in a way we can detect), we need to stop execution again with the same process.
1. The only way for us to know if agent supports telemetry, is to check if HandlerEnvironment.json file contains key 'eventsFolder' or 'eventsFolder_preview'. This check is done in ExtEnvHandler.py when it reads contents from HandlerEnvironment.json.
2. Agent will independently read events we write every 5 mins and there is not reporting mechanism on the VM for the status of these reads. Agent reports all issues in GuestAgentGenericLogs table. 
3. The one check I can think of here, in case an events folder is provided in HandlerEnvironment.json, validating the dir path and reporting a failure in case if the validation fails.


Like I mentioned earlier, the eventsFolder and eventsFolder_preview check is done in ExtEnvHandler. This reads contents of HandlerEnvironment.json and saves status folder, config folder, log folder and event folder paths. In case of events folder, it looks for 'eventsFolder' key within the file first, if not found, looks for 'eventsFolder_preview' key. If neither of these are found, saves None in the ext_env_handler.events_folder
This is then checked in telemetry setup in ActionHandler (here), validating the dir path and reporting whether telemetry is supported

Reply:
Meta-comment: These are things that need to go into a design doc and reviewed prior to implementation. Please ensure all of these are covered in a design doc and reviewed prior to further reviews on the full PR.

Specifically here: For *canary testing only*, we can just stop execution when the agent is known not to support telemetry and write it to the log file. This necessitates testing against a VM with an older agent version. The pre-requisite is the conversation that needs to happen with the agent team on how to best prep test environment faithfully for that.

Reply:
Updated the code to stop extension flow, logging an error in handler log